### ISSUE TITLE
es6-shim breaks Chrome App CSP

### ISSUE THREAD

**Issue by stewart** [paulmillr/es6-shim#301]
Unless sandboxed, `es6-shim` cannot run inside of Chrome Apps right now due to [this line](https://github.com/paulmillr/es6-shim/blob/master/es6-shim.js#L72), which triggers the CSP, and is functionally an `eval()` statement.

I'm unsure if the best fix is reverting https://github.com/paulmillr/es6-shim/commit/ceeb51cc99bbe3729e96eae8c7af966a2d67dfde, or if a better solution can be found.

**Comment by sandstrom** [comment]
CSP is quite handy and — in addition to being used in Chrome apps and Firefox OS — many web applications also use it. We're one of them and just ran into this.

Would anthonyryan1's method work? [https://github.com/paulmillr/es6-shim/issues/301#issuecomment-64966395]

**Comment by ljharb** [comment]
@sandstrom See my response right below it: https://github.com/paulmillr/es6-shim/issues/301#issuecomment-64973036

**Comment by stewart** [comment]
I think I've got a possible fix here - https://gist.github.com/stewart/321560330b152dfde649.

Didn't cause any issues in the environments I tested in, but maybe I'm missing something?

**Comment by ljharb** [comment]
@stewart What did you test it on? It would need to work back to IE 6, in node 0.6 and later, all io.js versions, node-webkit/nw.js, web workers, the node `vm` module's sandbox, and probably in all of a CJS, AMD, and no-module environment for each.

**Comment by stewart** [comment]
The one major drawback to my approach that the CSP-violating version doesn't have is that if `es6-shim` gets wrapped up in an IIFE w/ strict mode enabled (say by a minifier or project bundler), it'll stop working.

On the other hand, a situation like that is easier to work around for developers than a CSP-violating dependency, and relatively easy for the library to detect (check if `globals === undefined`).

Travis indicates at least Node/iojs support about on par with `master` - https://travis-ci.org/stewart/es6-shim/builds/69499282, I'll see about testing other platforms later (don't have access to IEs older than 10 though, I'm afraid). I did test in Chrome/Safari/Firefox and they all seemed to like it.

**Comment by ljharb** [comment]
I'm less concerned about browsers and node (where `window` and `global` are reliable enough), than I am about web workers and node-webkit, both of which we can't currently test on travis, nor do I have any way to test locally.

**Comment by stewart** [comment]
I don't have test harnesses available for either at present, but I'll see about getting something worked together over the weekend, unless someone else beats me to it.

**Comment by ljharb** [comment]
Thanks!

**Comment by ljharb** [comment]
I've worked on this with @rwaldron and @jugglinmike in person, and we've found a non-eval approach that *seems* to work in browsers, workers, node-webkit, and node.

In order to unblock Chrome apps from having CSP issues, I'm going to go ahead with this change. Please let me know if this change has broken anything.

**Comment by Yaffle** [comment]
@ljharb  this throws an error when running using WSH.
Altough, es6-shim does not work in WSH anyway...

**Comment by ljharb** [comment]
Thanks, that's good to know - what is the thrown error?

If you can provide me with various exceptions that it throws (as separate issues, please) I'll do my best to fix them!

**Comment by Yaffle** [comment]
`throw new Error('unable to locate global object');`

**Comment by ljharb** [comment]
Perfect, that's exactly what it's supposed to do :-) It appears that in WSH, the only option is to get `this` in the global context. http://stackoverflow.com/questions/14450424/equivalent-to-window-in-jscript-runtime

I'd suggest then that in wsh, users need to do something like `var global = this;` in the global context prior to including the `es6-shim`.

**Comment by Yaffle** [comment]
@ljharb , `setTimeout` is not supported, so
```
var promiseSupportsSubclassing = supportsSubclassing(globals.Promise, function(S) {
			return S.resolve(42).then(function() {}) instanceof S;
		});
```
`promiseSupportsSubclassing` test throws an error...

**Comment by ljharb** [comment]
@Yaffle Try the latest commit in WSH - please file a new issue for me if there's any further issues.

**Comment by Yaffle** [comment]
@ljharb , well... it works with `var global = this;` and es5-shim, but throws some errors:
1) https://github.com/paulmillr/es6-shim/blob/master/es6-shim.js#L1155
"Number" is a bad name for IE8 (probablye because of http://kiro.me/blog/nfe_dilemma.html), so before this line is reached, `Number` is called and throws the error, because `isBinary` is not defined yet.
2) `wrapConstructor` uses `Object.getOwnPropertyNames`, which is not supported in IE8 event with es5-shim;
3) `Map` and `Set` are not polyfilled in IE8 because of `if (supportsDescriptors) {`

**Comment by ljharb** [comment]
@Yaffle Can you please file an issue for 1 and 2? 3 is expected and desired.

**Comment by chmx** [comment]
@ljharb Using `var global = this;` before loading es6-shim works with Nashorn too. Thanks!
